lib/helper.js

```diff
@@ -166,3 +166,64 @@ exports.defer = function () {
     promise: promise
   }
 }
+
+/**
+ * Patches di module to support ES2015 classes.
```

Comment:
While I think it is cool that you worked this out, I am not keen on this strategy and this implementation. The karma-special "DI" system makes karma difficult to work with until one spends time decoding what it does. To me, this DI patching approach makes the problem worse: now we have another layer of confusion hiding the first confusion.  

The migration to ES classes will make the karma code easier to read for most developers.   Can we find a way to achieve that without the DI patching?

The karma DI works through instances exported and matched with some funky $inject coding. The conversion to classes changes source which is fundamentally singleton into class structure, which then requires object creation to fit back into the DI system.  Rather than hacking the DI system to create instances from exported classes, consider leaving the DI and module interfaces unchanged in this first pass.  Create instances in the new ES modules, export them just as karma does now. That gives us ES class support without obscuring the obscure. It also avoids bundling changes that will make bug tracking very difficult.   Then we can tackle the DI issue separately.

Reply:
Well, after reading the source code of `di` module I find it very straightforward and simple to understand DI implementation 😄 And the fact that it worked all this time without breaking only confirms the fact. Not sure why users are having difficulties working with it, maybe a bit of documentation can completely resolve the problem.

I doubt there is a much better/more type safe dependency injection mechanism possible using plain JS at the moment.

> Can we find a way to achieve that without the DI patching?

I think I can convert all `instantiate()` calls and `type` providers to use `invoke()` and factory methods instead and it will work without a DI patch. I'll check it later today. 

Something like [here](https://github.com/karma-runner/karma/pull/2948/files#diff-c945a46d13b34fcaff544d966cffcabaL102):
```
- capturedBrowsers: ['type', BrowserCollection],
+ capturedBrowsers: ['factory', function (emitter) { return new BrowserCollection(emitter) }],
     args: ['value', {}],
```

Reply:
Documentation would be awesome. 

I agree that the DI system has proven useful. Nevertheless it is unusual and unique, requiring users to read the code for a probably unrelated subsystem to make progress on changes. For your change however, the issue is only to not make things more complex if we can achieve your goal within the current system.

Reply:
I've removed a `di` patch and replaced with factory functions.